### ISSUE TITLE
version 2.x refactoring to support Multiple Grid 

### ISSUE THREAD

**Issue by ghiscoding** [ghiscoding/aurelia-slickgrid#19]
Here's a list of what cannot be done currently in version `1.x` until a new major release comes out `2.x` which mean breaking changes
- [x] 1. GridEvent Service, no longer references the 2nd argument `gridOptions`  (closed by PR #51)
- [x] 2. a lot of the Services instantiated in `aurelia-slickgrid` ViewModel use an `init()` which usually ask for `init(grid, gridOptions, dataView)` but once we get the grid object, we actually have access to the `gridOptions`, so we can simplify this a lot (closed by PR #55).
- [x] 3. remove all deprecated code (just search for the word deprecated), code change could be renamed or re-implemented
  - [x] a. `onBackendEventApi` was re-implemented and replaced by `backendServiceApi` ([commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/05568ad1524e8c8dfcf2801fabb746c369463adb))
- [x] 4. remove possibly unnecessary global Event Aggregators (EA) by Element Dispatch (ED)
  - [x] ~a. to be discussed, having EA globally might help in some cases, in my current project I have a left side menu that when triggered needs to resize the grid (autoResize doesn't work in that case) and I don't think it would be possible without an EA.~
   - [x] b. use `asg` for Aurelia Events and `sg` for SlickGrid/DataView Events & rename `eventPrefix` to `slickgridEventPrefix` ([commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/8eb1ad493eeeab18f3fa5943dde6ca1bab6972d2))
- [x] 5. Filter could only use `searchTerms` array instead of doing multiple logic to deal with `searchTerm` (singular) and `searchTerms` array ([commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/686885a5439fe9b1e6bd9a78993d7edb0d049425)).
- [x] 6. Remove all previous reference of `selectOptions` in all Filters ([commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/532554dc5dd5111867ebb4c9746ff5b9fe94b3ff))
- [x] 7. Editors/Filters should use a common `Column` property and possibly remove the `filter` Column property. 
- [x] 8. remove `FormElementType` from Filters, it should **only** use `FilterType` ([commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/895474a85aba3c2dc0c74134065e36818121b560))
  - [x] a. the `FormElementType` is actually not used anywhere else I think, so might be better to delete it completely
- [x] 9. BackendService API `onFilterChanged` Event and function should not be named the same, it is rather confusing when troubleshooting, maybe change function to `processFilterChanged` ([commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/0bf9d940a92d1045cad9c78412fda5183f9687c3))
- [x] 10. remove `exportWithFormatter` from GridOptions.interface since it is now part of `exportOptions` ([commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/bd83f943e584226383bb0325096d2206972bcd04))
- [x] 11. GridMenu has both `hideX` and `showX` flags, they should all be `hideX` since that is the default name in the external plugins ([commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/aa32a56de1d1657bf74664c55dc8c8cf73929d07))
- [x] 12. Possibly rename the `GridExtraService` to simply `GridService` ([commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/d81b37ba205922afeb07148de586591432887919))
- [x] 13. A lot of the Editors properties (like `collection`) are available only from the generic `params` property which removes the benefit of intellisense and type check. This goes with number 7.
- [x] 14. Remove the `initOptions` from the `BackendService`, that was replaced by `init` some time ago ([commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/6e5927cea0b12239bc9063a3d7b3175903f29d47))
- [x] 15. `GridExtraUtil` contain only 1 function, it would be better to move it into the renamed `GridService` and delete `GridExtraUtil` completely ([commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/4e31b1896c7501e2f6af059d0c4c2b447815f02a))
- [x] 16. Should pass `i18n` as a Grid Options instead of a generic `params` grid option and fix a column picker issue ([commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/d2b17919c8491261cc17f6e4b5b9f65e47db6ee3))
- [x] 17. change all Singleton Services to non-Singleton ([commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/daf98c16f7e0631c6f67ab81bf801a9bdbe3b3f9))
- [x] 18. renamed onGridStateServiceChanged to onGridStateChanged
- [x] 19. add multiple grids in a view ([commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/27d1d2b6b8821250b0597079be14b01cc5a14e8d) and [commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/e9306b363c0ae833dfdb246bb57af87000e9d462))
- [x] 20. change `column.interface` signature of `onCellClick` and `onCellChange` to expose event as first argument to keep consistent with SlickGrid, see issue #60 
- [x] 21. update [Migration Guide](/ghiscoding/aurelia-slickgrid/wiki/Migration-to-2.x)
- [x] 22. update all `client-cli` and GitHub `demo` samples
- [x] 23. update Wikis and there's a **lot**

This list will be taken into consideration in the version `2.x` which will be a Major and breaking version.
Simply keeping this for reference as of now. 

Also, don't expect a version `2.x` any time soon. 
This will be addressed only when we decided that all features that we want are in place and we want to get rid of all properties and naming that weren't of good use in `1.x`.

**Comment by ghiscoding** [comment]
Here's a list of what cannot be done currently in version `1.x` until a new major release comes out `2.x` which mean breaking changes
- remove all deprecated code (just search for the word deprecated) because it was renamed or re-implemented
  - `onBackendEventApi` was replaced by `backendServiceApi`
- remove possibly unnecessary global Event Aggregators by Element Dispatch
  - to be discussed, have EA globally might help in some cases

That's all I can think of right now, expect the list to grow. 
Simply keeping this for references.

**Comment by ghiscoding** [comment]
Migration Changes Guide from version `1.x` to `2.x` 
- `GridExtraService` got renamed to `GridService`
- `GroupingAndColspanService` got renamed to `GroupingService`
- `ControlAndPluginService` got renamed to `PluginService`
   - **not finalized** this might end up being split into 2 Services
- For `BackendServiceApi`, the `service` property now as to contain a `new` instance of the Backend Service that you want to use (`GraphqlService` or `GridOdataService`) like so
##### Before
```javascript
export class MyGrid {
  constructor(private graphqlService: GraphqlService, private i18n: I18N) {
  }
    this.gridOptions = {
      backendServiceApi: {
        service: this.graphqlService,
        preProcess: () => this.displaySpinner(true),
        process: (query) => this.getCustomerApiCall(query),
        postProcess: (result: GraphqlResult) => this.displaySpinner(false)
      }
    };
```
##### After

```diff
export class MyGrid {
-  constructor(private graphqlService: GraphqlService, private i18n: I18N) {
+  constructor(private i18n: I18N) {
  }
    this.gridOptions = {
      backendServiceApi: {
-        service: this.graphqlService,
+        service: new GraphqlService(this.i18n),
        preProcess: () => this.displaySpinner(true),
        process: (query) => this.getCustomerApiCall(query),
        postProcess: (result: GraphqlResult) => this.displaySpinner(false)
      }
    };
```

**Comment by ghiscoding** [comment]
@jmzagorski 
How do you want to address number 13. ?

**Comment by jmzagorski** [comment]
For number 7 i think you wanted to keep  the editor and filter configuration separated in the `columnDef`. Probably a good idea since both accept a `collectionFilterBy` property and I have a grid that dynamically filters the `params.collection` by changing the `columnDef.params.collectionFilterBy` on cell click, while leaving the `filter.collection` alone. Not sure if that can be accomplished if both the editor and filter use the same configuration?

That would mean leave the editor on the params or move them to an editor specific property (in reference to number 13)

**Comment by ghiscoding** [comment]
@jmzagorski 
I would like to put some Types for Editors, not just for Filters. The problematic is even though we can use `filter`, we can't use `editor` since it is used internally by SlickGrid itself. We could call it `editorOptions` though, and if so, we could potentially rename `filter` to `filterOptions` to make them equal (however that would create more refactoring in the migration, but I'm ok with that). 

The end goal is to get some Types (and intellisense when the coding editor supports it, like VScode). I wanted to wait until Filters and Editors were stable enough, I think they are now, and for version `2.x` before revisiting them. I think we have everything we need to make a decision, which I kinda need rather soon because I already started the Major version work in my Angular repo. 

I don't think that leaving all of the Editors stuff in `params` is a good idea, especially if we want more Typing.

**Comment by jmzagorski** [comment]
Here are a few other options I could think of to brainstorm about:

1. Instead of `filterOptions`, `editiorOptions` etc, make a property called `options` and then put `filter` and `editor` as properties (`columnDef.options.filter`, `columndDef.options.editor`)

2. Make your public API `columnDef.editor` a `Editor | EditorOptions` so it is similar to the filter. Then in `AureliaSlickGridCustomElement` move the editor option properties to another property on the `columnDef` and put the actual `Editor` interface on the original `columnDef.editor` for slickgrid to use.

**Comment by ghiscoding** [comment]
I was thinking about your number 1. too, the downside is that it makes the property calls longer. Unless you want to merge Editor/Filter properties and only separate the ones that you want? But that might be confusing for collection stuff

This kinda mean that Editor and Filter will have the exact same collection behavior. Is that **always** wanted?
```javascript
options: {
       collection: [ 
         { value: '', label: '' }, 
         { value: true, label: 'true' }, 
         { value: false, label: 'false' },
         { value: undefined, label: 'undefined' }
       ],
       collectionFilterBy: {
          property: 'effortDriven',
          operator: OperatorType.notEqual,
          value: undefined
       },
       collectionSortBy: {
          property: 'effortDriven',    // will sort by translated value since "enableTranslateLabel" is true
          sortDesc: false,             // defaults to "false" when not provided
          fieldType: FieldType.boolean // defaults to FieldType.string when not provided
       }, 
       type: FilterType.multipleSelect
   }
```

Your number 2. I'm not sure to follow you on how to fake `editor`. Are you saying to replace `editor` before passing the column definitions to the `new Slick.Grid`? I feel that it's kind of a patch

**Comment by jmzagorski** [comment]
I think combing it might be troublesome, but that is just a guess. For example, in my app I dynamically filter the `singleSelectEditor` when a cell is clicked, but the header filter still shows all the values from the same collection. I do this because each row can only have a subset of values from the entire collection based on another value in the same row. For example:

```javascript
    onCellClick: ({ columnDef, dataContext }) => {
      columnDef.params.collectionFilterBy = {
        property: 'jobTitleId',
        value: dataContext.ownerJobTitleIds,
        operator: OperatorType.contains
      };
    },
```

So if this was changed every time a cell was clicked, will this affect the header filter? I know the header filter is rendered early on so it might not affect it unless the header filter updates while the user is using the grid.

As for number 2, yes it is sort of "hacky" i guess. The only thing it does is makes your public API cleaner and not tied to slickgrid anymore. Here is a brief example of how it could work without changing the underlying column definitions:


```typescript
    this.grid = new Slick.Grid(`#${this.gridId}`,
       this.dataview,
       // if we made the Editor class property name "slickEditor"
       this._columnDefinitions.map(c => ({ ...c, editor: c.editor.slickEditor })),
       this.gridOptions);
```

**Comment by ghiscoding** [comment]
hmm maybe with the spread operator it doesn't look too bad. What is `c.editor.slickEditor`? I mean where does that come from? Aren't we just supposed to pull whatever the user added in `editor` and empty it out afterward we're done keeping it in a temp area that will be used by our EditorService?

**Comment by jmzagorski** [comment]
`c.editor.slickEditor` would be what `columnDef.editor` is today, i just gave it a random name so probabaly should use another name if we decided to use this method. For example:

**Today**
```javascript
{
    editor: Editors.date,
    field: 'due',
    filter: {
      type: FilterType.compoundDate
    },
    filterable: true,
    formatter: Formatters.dateIso,
    id: 'due',
    name: 'Due',
    sortable: true,
    type: FieldType.dateIso
  }
```

**Braistorm**
```javascript
{
    editor: {
      slickEditor: Editors.date,
     // ... other editor options here
    },
    field: 'due',
    filter: {
      type: FilterType.compoundDate
    },
    filterable: true,
    formatter: Formatters.dateIso,
    id: 'due',
    name: 'Due',
    sortable: true,
    type: FieldType.dateIso
  }
```

**Comment by ghiscoding** [comment]
but where is the part to swap the property in a temp variable? Isn't your `.map` supposed to be reassigned to column definitions after you're done? SlickGrid uses `editor` for it's Editor Factory, so we need to empty it out **before** creating the grid itself. If we use `editor` that can be renamed to a temp variable (let say `editorOptions`), isn't it suppose to look like the following?
```javascript
this.grid = new Slick.Grid(`#${this.gridId}`,
       this.dataview,
       // if we made the Editor class property name "slickEditor"
       this._columnDefinitions = this._columnDefinitions.map(c => ({ ...c, editor: { }, editorOptions: c.editor })),
       this.gridOptions);
```

BTW, I'd like to have the brainstorm done by the end of this week and implement it in the weekend and then in my Angular repo on Monday. I need multiple grids in my project, like **now** lol

**Comment by jmzagorski** [comment]
oh yeah i forgot that part, and I forgot about the editorFactory property on `gridOptions`. So we can blank out the editor and put our factory on `gridOptions.editorFactory.getEditor` or reassign the editor and editorOptions on the fly 

If you were to use editorOptions the reassignment would look like this:
```javascript
this.grid = new Slick.Grid(`#${this.gridId}`,
       this.dataview,
       // if we made the Editor class property name "slickEditor"
       this._columnDefinitions = this._columnDefinitions.map(c => ({ ...c, editor: c.editor.slickEditor, editorOptions: { ...c.editor } })),
       this.gridOptions);
```

If you were to use editor factory it would look like this:
```diff
-for (const c of this._columnDefinitions) {
-      if (c.editor) {
-        c.editor = Factory.of(c.editor).get(this.container);
-      }
-}
+ this.gridOptions.editorFactory {
+  getEditor: column => Factory.of(this.columnDefinitons.find(c => c.id === +column.id).editor.slickEditor).get(this.container)
+}

this.grid = new Slick.Grid(`#${this.gridId}`,
       this.dataview,
       // if we made the Editor class property name "slickEditor"
+       this._columnDefinitions = this._columnDefinitions.map(c => ({ ...c, editor: null, editorOptions: { ...c.editor } })),
       this.gridOptions);
```

**Comment by ghiscoding** [comment]
When I said Editor Factory, I really meant SlickGrid internal factory which is the column `editor` property, you can see it used [here](https://github.com/6pac/SlickGrid/blob/master/slick.grid.js#L1646) by the `getEditor()` function. But yeah we also have our own Editor Factory, which you did an excellent job at implementing it :) 

The `editor: c.editor.slickEditor` still confuses me, I don't know what it does since I believe the `editor` should be left empty or `null` as you pointed out in 2nd sample. 
BTW, `editorFactory` is also column property used internally by SlickGrid [here](https://github.com/6pac/SlickGrid/blob/master/slick.grid.js#L89), so we can't use that option either :P

Which one do you think would be the best option to go with?

**Comment by ghiscoding** [comment]
Dohhhh I feel like Omer Simpson now... I didn't realized that we do use the `editor` today lol
I guess we use it that way because that is what SlickGrid likes to see. Wow I'm sorry, I was missing an important piece since the start of the subject, my bad. I'm glad we talked because I would have missed it completely, now I understand the `editor: c.editor.slickEditor` it totally make sense

**Comment by ghiscoding** [comment]
@jmzagorski 
I want to make sure that you see number 4.b and 4.c from the [list](https://github.com/ghiscoding/aurelia-slickgrid/issues/19#issuecomment-369430250) on top. I would like to make a better separation of the dispatch events of SlickGrid versus the Aurelia-Slickgrid ones. Basically, I want SlickGrid to start with prefix `sg`, while the Aurelia-Slickgrid events to start with `asg`. This would help the user and us if there's any troubleshooting to do, to understand where the event is processed. Make sense?

**Comment by jmzagorski** [comment]
Makes sense yes. What does your html view on your left-side menu look like? You might be able to catch the event in your view if there is a parent element to your grid without using the EA.

**Comment by ghiscoding** [comment]
Ah the example that was written is with my current App, however it's almost identical to our demo with left-side menu, the only difference is that the left-side menu is collapsible. I can re-code our demo to look similar and add the collapsible. The idea is that the left-side menu and the right-side are containers, they don't know what goes in, neither do they have a relationship of Parent/Child, they are independent. 

See the print screen below, the "x" is clickable which will collapse and only show the icons and not the text. When that "x" is clicked, it doesn't fire an DOM resize events, I think it's due to the fact that I only use a CSS property change and not an actual element resize... all that to say, I had to go with a Global Event Emitter (which in the Aurelia is an Event Aggregator)

![slickgrid_admin_menu](https://user-images.githubusercontent.com/643976/40442099-d04f6c40-5e90-11e8-9cc2-d8c73da98961.png)

**Comment by ghiscoding** [comment]
@jmzagorski 
I am almost done with the refactoring of `editor`, I decided to go with the following
```javascript
this._columnDefinitions = this._columnDefinitions.map((c: Column | any) => ({ ...c, editor: this.findEditor((c.editor && c.editor.type), c), internalColumnEditor: { ...c.editor } })),
```
Some notes
1. To make the `editor` to behave exactly like `filter`, I created a new interface `EditorType` which can be used inside the `editor.type` (exactly like `filter`).
2. instead of using `editorOptions` to internally swap the options, I decided to use a more familiar name `internalColumnEditor` which tells the user to not use that property
3. the function `findEditor` will be a Factory call in our Aurelia-Slickgrid, I couldn't find how to create a factory in Angular and so I just did a quick and dirty collection find instead of a factory.
4. the `editor` is of type `ColumnEditor` (just like `filter` was of type `ColumnFilter`) and this brings all the usual properties that were previously used the generic `params`
5. with the new `EditorType`, I was able to easily add a `custom` and created a Custom Editor and it... sweet
   - however, I can't `new` it since it has to a valid SlickGrid class (not an instance) because if I do, I won't get access to the `args` required in the constructor (provided internally by SlickGrid)
6. **very important** there is only 1 issue that stands, the `CollectionFormatter` was using the Collection from the generic `params` and now it's no longer available. I can take it from the `editor` but wait I can't... why? because I don't know if the Collection is inside the `editor` or the `filter` object!?! 
   - it's also not just the collection but also `customStructure` and possibly more in the future, 
   - so for now, I left it inside `params` for the Example 3 (Editor) to still work.
   - I need your input for this one

The commit for all of that in the Angular-Slickgrid repo can be seen [here](https://github.com/ghiscoding/Angular-Slickgrid/commit/bd8440b5d50ab8338b5b32e3d8a68643dda75d12) and I will release a beta version in that repo in the coming days

**Comment by jmzagorski** [comment]
I am pretty sure `CollectionFormatter` should be from the editor because that function formats values selected from the editor.

> however, I can't new it since it has to a valid SlickGrid class (not an instance) because if I do, I won't get access to the args required in the constructor (provided internally by SlickGrid)

Not sure what you mean by this. Are you newing up the editor class yourself?

**Comment by jmzagorski** [comment]
okay I see what you mean with the `CollectionEditorFormatter`. If the user does not have an editor, they still might need to use the `CollectionFormatter`.

What about the `CollectionFormatter` checking if there is an editor then using the `editor.collection` property and if not fall back to the `params.collection`?

**Comment by ghiscoding** [comment]
@jmzagorski 
I was thinking about the fallback but then if the user also has the collection inside the `filter`, then what? which one has priority and such? Also the `if` condition is gonna be huge, take a look at the `if` [code](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/formatters/collectionFormatter.ts#L9) now. I would prefer to simply create multiple formatters, it's easy to create and doesn't add much more code, it's also more focused and less confusing

**Comment by jmzagorski** [comment]
Right now the `CollectionFormatter` ignores the `filter.collection` and always uses the `params.collection` which is only used by the `editor`, so would we have to worry about `filter.collection`?

**Comment by ghiscoding** [comment]
@jmzagorski 
The thing is that the `collection` is now inside the `editor` object, no longer inside `params`. I think you knew that already, but just making sure you had that info.

A formatter is meant to format the data, so there's nothing anywhere that says it is meant to be used in combo with the Editor. Having a new `CollectionEditorFormatter` is much more explicit, and we can leae the `CollectionFormatter` with `params`, in case you want to use it without any `editor` or `filter`.

**Comment by jmzagorski** [comment]
okay. I see what you are doing. If you did not want to create a formatter every time there is a collection under a different column property, you could create a factory function

```typescript
export const collectionFormatter: Formatter = (columnProperty) => (row: number, cell: number, value: any, columnDef: Column, dataContext: any) => {
  if (!value || !columnDef || !columnDef[columnProperty] || !columnDef[columnProperty].collection
    || !columnDef[columnProperty].collection.length) {
    return '';
  }

  // etc. replace every params with [columnProperty] which can be filter or editor or whatever else
};
```

View model would be 
```javascript
const columnDef = [{
  formatter: Formatters.collection('editor')
}]
```

Just an idea, you dont have to use it.

**Comment by ghiscoding** [comment]
oh that would be nice indeed, but TS seems to not like the way you wrote. 

![image](https://user-images.githubusercontent.com/643976/40683226-edc8b784-635b-11e8-99f1-b8f87c4d8f16.png)

**Comment by jmzagorski** [comment]
oh yeah, because the `columnDef.formatter` will return a `Formatter` or a `Func<string, Formatter>`

**Comment by ghiscoding** [comment]
I'll leave it as a 2 formatters for now (`CollectionFormatter` and `CollectionEditorFormatter`), I'll let you play with that before the release. I would say this is minor and can wait.

**Comment by ghiscoding** [comment]
@jmzagorski 
Sorry I'm going to have to bother you again, once this is done, I think the Major Version will be ready for a Beta spin.

So I was working late on implementing the `editor` with the same code as what I have in my Angular repo. Pushed a [commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/0ca7bf84c56c3ebec2e4e8abd3766f296ed5a5cf). But I'm having some issues with this piece of code at this [line](https://github.com/ghiscoding/aurelia-slickgrid/blob/0ca7bf84c56c3ebec2e4e8abd3766f296ed5a5cf/aurelia-slickgrid/src/aurelia-slickgrid/aurelia-slickgrid.ts#L143)
```javascript
this._columnDefinitions = this.columnDefinitions.map((c: Column | any) => ({ ...c, editor: this.getEditor((c.editor && c.editor.type), c), internalColumnEditor: { ...c.editor } }));
``` 
fails because `c.editor` is not resolved yet I believe because of you `Factory.of` loop that you've put in the `bind` [here](https://github.com/ghiscoding/aurelia-slickgrid/blob/0ca7bf84c56c3ebec2e4e8abd3766f296ed5a5cf/aurelia-slickgrid/src/aurelia-slickgrid/aurelia-slickgrid.ts#L262),

Also, I just wanted to get it going but eventually I want to replace the `getEditor()` function [here](https://github.com/ghiscoding/aurelia-slickgrid/blob/0ca7bf84c56c3ebec2e4e8abd3766f296ed5a5cf/aurelia-slickgrid/src/aurelia-slickgrid/aurelia-slickgrid.ts#L295) by a Factory function. I didn't find how to do that in Angular and didn't want to spend time, but that should be easy to do in Aurelia. Once that is done the [AVAILABLE_EDITORS](https://github.com/ghiscoding/aurelia-slickgrid/blob/0ca7bf84c56c3ebec2e4e8abd3766f296ed5a5cf/aurelia-slickgrid/src/aurelia-slickgrid/editors/index.ts#L26) can go away.

Again this is the last stretch, I've put all the fixes and changes that I've done from my Angular repo to here. we are getting close... :)

**Comment by ghiscoding** [comment]
Actually thought about it, we just have to move the reassignment of the column definition into the `bind()` function and also we can use the `Factory.of` right from there. However there is an exception, the `EditorType.custom` won't work since it's not part of the factory and has to be loaded with `column.editor['customEditor']`

If you have the chance to look into that during the day, that would be great.

**Comment by jmzagorski** [comment]
Was there a reason why you made the `columnDef.editor` interface a `editor?: EditorType | ColumnEditor;`  instead of a `editor?: Editor | ColumnEditor` and then the `ColumnEditor.type` property would be the editor class passed by the user as we do it in version 1? Then you dont need a factory since they are passing us the class

**Comment by ghiscoding** [comment]
right that is probably why I need to make it a type `any` somewhere in my code. it is a mistake and yes it should be `editor?: Editor | ColumnEditor`

As for the type, it wasn't working with the type being something else than enum. Also, the advantage of enum is that it brings the custom editor as a new type. I wanted to make the `editor` have the same structure as `filter` and by using an enum, it is the same structure. You can try with type as `Editor` but when I tried it was failing on me.

**Comment by jmzagorski** [comment]
I was just curious.

Is there any benefit to using an enum over the actual class being passed to `type`? Looking at the custom filter implementation:  

```javascript
filter: {
       type: FilterType.custom,
       customFilter: new CustomInputFilter() // create a new instance to make each Filter independent from each other
    }
```

It makes the filter interface more verbose and also makes more work in your library to instantiate the one they want to use based on an enum when the users could just pass either your filter or their custom one:

```diff
filter: {
+     type: Filters.select // or new CustomInputFilter()
-       type: FilterType.custom,
-       customFilter: new CustomInputFilter() // create a new instance to make each Filter independent from each other
    }
```

Just a thought

**Comment by ghiscoding** [comment]
I tried changing the type as you said and it works, here the [commit](https://github.com/ghiscoding/Angular-Slickgrid/commit/24ce5dbda8197523c3bb1c21f0c04e2612aaa35f) in Angular repo. I can't work on the Aurelia one until after work.

The only benefit in using the enum is that it doesn't pass the entire class when you import `EditorType` versus the `Filters` (basically if we do `Editors.checkbox`, like in version 1). I don't mind changing both the Filter and Editor if it helps to make the library lighter and less complex, will the DI still work though? For the Custom Editor, I couldn't do the `new` since it was conflicting with SlickGrid as I wrote [here](/ghiscoding/aurelia-slickgrid/issues/19#issuecomment-392854423)

**Comment by jmzagorski** [comment]
the DI should still work. We would change the inner workings of the `filterFactory` and just use `this.container.get(columnDef.filter.type)` since the `filter.type` will be a class export instead of a name. That would remove the need for the user to remember to register their custom filter as a plugin for this library and there would be no need for a `filterType` property on the class.

**Comment by ghiscoding** [comment]
sounds good, if you have a chance to do, it would help out a lot. Else I might have few questions to get it going

**Comment by jmzagorski** [comment]
okay, I will see if it even works or makes things simpler

**Comment by ghiscoding** [comment]
@jmzagorski 
so I tried to move the column definition reassignment into the `bind()` function and that worked. Here's the code without the comments. 

```javascript
bind() {
  this.gridOptions = { ...GlobalGridOptions, ...this.gridOptions };
  this._columnDefinitions = this.columnDefinitions.map((c: Column | any) => ({ ...c, editor: this.getEditor((c.editor && c.editor.type), c), internalColumnEditor: { ...c.editor } }));

  this.columnDefSubscriber = this.bindingEngine.collectionObserver(this.columnDefinitions)
    .subscribe(changes => this.updateColumnDefinitionsList(this._columnDefinitions));

  for (const c of this._columnDefinitions) {
    if (c.editor) {
      c.editor = Factory.of(c.editor).get(this.container);
    }
  }
}
```
Also I made a mistake in the `sg-on-cell-changed.delegate` it should be `change` not `changed`, that is to show the editor changes. 

I **did not** commit any changes, I only wrote it here for reference. I also thought you might be in the middle of playing with the code on your side as well. Any luck with the refactoring you mentioned earlier? I'll let you push your changes tomorrow (Friday), it would be really nice to get a Beta release by early next week.

**Comment by jmzagorski** [comment]
Also, you can simplify the bind if you want:

```diff
bind() {
  this.gridOptions = { ...GlobalGridOptions, ...this.gridOptions };
  this._columnDefinitions = this.columnDefinitions.map((c: Column | any) => ({ 
    ...c,
-    editor: this.getEditor((c.editor && c.editor.type), c),
+   editor: c.editor && Factory.of(this.getEditor(c.editor.type), c)).get(this.container),
    internalColumnEditor: { ...c.editor } 
  }));

  this.columnDefSubscriber = this.bindingEngine.collectionObserver(this.columnDefinitions)
    .subscribe(changes => this.updateColumnDefinitionsList(this._columnDefinitions));

-  for (const c of this._columnDefinitions) {
-    if (c.editor) {
-      c.editor = Factory.of(c.editor).get(this.container);
-    }
-  }
}
```

**Comment by jmzagorski** [comment]
Also do you think `internalColumnEditor` is a little verbose since the property is in the `columnDef` object (as opposed to just `interalEditor`)? (lol I know I analyze everything)

**Comment by ghiscoding** [comment]
the name is of the type that goes in, it is of type ColumnEditor, so I just added internal in the front :P 
I don't mind changing the name, it's not to be used outside anyway

As for all the Editors imported, it's most probably a mistake from the first draft to get it working. I don't think we need them anymore, you can delete them

**Comment by jmzagorski** [comment]
ok. As for the name i think `internal` is good, I was talking about `column` in `internalColumnEditor`. Do you think it would be better naming it `internalEditor` instead of `internalColumnEditor` since the property is on the column object already.

**Comment by ghiscoding** [comment]
I don't mind changing the name if it bothers you, if you are in the middle of code change, can you make sure that these changes displayed in this Angular repo  [commit](https://github.com/ghiscoding/Angular-Slickgrid/commit/24ce5dbda8197523c3bb1c21f0c04e2612aaa35f) are also part of our Aurelia lib. Thanks

**Comment by jmzagorski** [comment]
it does not bother me since it is internal, i just thought I would bring it up for discussion to see if it sounded redundant or not

**Comment by jmzagorski** [comment]
On the `ColumnFilter` interface there is an `operator` property and in your `filter.service` you are selecting the `operator` based on the `FilterType` property [here](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/services/filter.service.ts#L197) for non-custom filters.

Since I am trying to make the `Editor` and `Filter` interface match (where both `type` properties are the class and not enum), this switch logic is problematic. 

We could remove the switch logic and encapsulate the `operator` value as a read-only property in the `Filter` interface since the operator should always be the same according to your comment [here](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/services/filter.service.ts#L197) (e.g the `SingleSelectFilter` will have a property called `operator` and that will have a value of 'EQ'. )

Not sure if that is what you want, but it is one way to reduce code in this service and add more encapsulation if the filter's operator will always be the same publicly. I see you do use 'IN' [here](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/filters/singleSelectFilter.ts#L109), but that seems more internal to clear out values since you do not want to use 'IN' when actually filtering a `SingleSelect`. If the `operator` is on the interface, then custom filters have to put it there so you can access the operator on any filter using `filter.operator`

However I am not that familiar with `operator` elsewhere.

**Comment by ghiscoding** [comment]
The operator cannot be read only, it is often set be 1 operator for a lot of Filters but now all of them. It is passed when using Compound filters and can also be changed when using a string filter (you can write `*Doe` and that will return every string that finishes with `Doe`. We could also add more Compound filters in the future, for example I'd like to have `CompoundMultipleSelect` (which would have an `IN` and `NIN` as operator you can choose.

Here are the possible Operators showing for a Compound Date Filter

![compound](https://user-images.githubusercontent.com/643976/40848740-0d02997e-658e-11e8-80e1-e558a5eb7900.png)

**Comment by jmzagorski** [comment]
ahh that's right. Do you have any suggestions for that switch logic in the `FilterService` if `Filter.type` is changed from `FilterType` to `Filter`? You could possibly run a switch statement on the class signature, for example:

```diff
        switch (filterType) {
-          case FilterType.select:
+         case Filters.select:
-          case FilterType.multipleSelect:
+         case Filters.multipleSelect:
            operator = 'IN';
            break;
-          case FilterType.singleSelect:
+         case Filters.singleSelect:
            operator = 'EQ';
            break;
          default:
            operator = operator;
            break;
        }
```

However that is assuming the `Filter.type` is a class signature and not an instance. This looks like the last hurdle in changing the `type` property on both the `Editor` and `Filter`

**Comment by ghiscoding** [comment]
isn't it better to simply add a default operator type inside each Filters? 

For the code you wrote, can we actually do a switch case on a class? what does `typeof Filters.select` return?

**Comment by jmzagorski** [comment]
The `typeof` would return `Function` for all of them so I dont think that would work.

The `defaultOperator` would work since that logic is only for an operator that is not already set, and you would not have to have the addition logic to check if it is a custom filter because you would be able to handle custom default operator types:

```diff
+     operator = operator ? operator : columnDef.filter.defaultOperator;
-     const filterType = (columnDef.filter && columnDef.filter.type) ? columnDef.filter.type : FilterType.input;
-     if (!operator && filterType !== FilterType.custom) {
-       switch (filterType) {
-         case FilterType.select:
-         case FilterType.multipleSelect:
-           operator = 'IN';
-           break;
-         case FilterType.singleSelect:
-           operator = 'EQ';
-           break;
-         default:
-           operator = operator;
-           break;
-       }
-     }
```

Not sure if you like that? Then we could get rid of the FilterType all together. I see you do add a `type` to `_columnFilters` array in this class, but I don't see it used anywhere?

Also I think `editor?: Editor | ColumnEditor` should be `editor?: any | ColumnEditor` right? Because `Editors.longText`, `Editors.text` are the class exports and not an instance. I say that because i get compile errors in example 11 when I changed type to `type: Editors.longText`. Maybe I am doing it wrong.

**Comment by ghiscoding** [comment]
The operator seems good the way you wrote it, however you can simplify it a bit. I mean code wise, the result will be the same, but you can write it this way too 
```diff
-     operator = operator ? operator : columnDef.filter.defaultOperator;
+     operator = operator || columnDef.filter.defaultOperator;
```
Well actually, you can only write it that way if  `operator` is never `0` and never `false`, else you have to use the ternary operator.

The `editor?: Editor | ColumnEditor` is working in my Angular repo but if you are in the middle of getting rid of `EditorType` and going with `Editors`, then I guess you'll have to use `any`. I'm not sure why but TS doesn't seem to like that we put the type `Editor`, technically it should work since `LongTextEditor` implements `Editor` or is extending, I forgot.

**Comment by jmzagorski** [comment]
I think TS is expecting an instance on the `type` property if it is `editor?: Editor | ColumnEditor` right? and `Editors.longText` is not an instance of `Editor`, but is a class export because slickgrid is creating the instance for us.

**Comment by ghiscoding** [comment]
Yeah it's probably something like it, too bad that we have to put `any` to go around the issue though.

Are you gonna be able to push your code by the end of the day? 
Hey I also meant to ask before but kinda forgot, how did your presentation go with the grid? 
Are they satisfied with it, is it missing anything feature wise?

**Comment by jmzagorski** [comment]
Yes, I will try to get it in tonight, if not tomorrow. I will probably break it up into 2 commits (editor then filter)

The presentation went well, they like the grid. The only feature I know about is the ability to show/hide their columns and save that configuration (I can't remember if you implemented that. If you did I have not had time to put it in my app). Basically I have a Dashboard page that has your grid and they can choose from a bunch of different columns so they can hide/show which they want). Since the app is so new they might have more requests once it is in production. ill add some screen shots and explain a little more how I am using the grid in that documentation issue.

**Comment by ghiscoding** [comment]
funny, I'm actually looking at this right now (visibility) lol
BTW it's not implemented, I opened this issue #67 to cover it

**Comment by jmzagorski** [comment]
okay i pushed the editor change. I may get to the Filter tonight or tomorrow (i am dealing with a webpack+DllPlugin+karma+aurelia issue for my unit tests ). I'll let you review that commit in case you don't like the direction that change went.

**Comment by ghiscoding** [comment]
I reviewed quickly your last commit and I can see that it was exactly the first approach that I went **before** trying to make the structure look like Filter. 

I guess it's ok for the new path, however I think that `type` is no longer a good word and I don't want previous user to think it is an `EditorType` or `FilterType` since it's no longer a type per say. So can we have another property name, perhaps something like `class`? or another word you think is better, but not `type`, especially for Filters if you are to do the same structure as this new Editors

**Comment by jmzagorski** [comment]
hmmm, naming...it is the hardest part of coding. I would say class, but that is very specific to es6 since technically someone can use a regular old function, and it is not an "instance" of a class either. It is really the constructor/named export

**Comment by ghiscoding** [comment]
The new standard is to not use an "I" in the front of an interface anymore, so I wouldn't use that. As for `interface`, I'm not sure this is good either since it's more of a class that is instantiated while an interface is just a structure per say.

**Comment by jmzagorski** [comment]
> The new standard is to not use an "I" in the front of an interface anymore

That is what I get for not keeping up with typescript! Thanks for updating me.

i thought about class, but the user is passing the library a non-instantiated class because Slickgrid is instantiating it for us [here](https://github.com/6pac/SlickGrid/blob/master/slick.grid.js#L3089), which would make the property on our editor column a constructor function. Maybe that is getting too technical

**Comment by ghiscoding** [comment]
How about `classReference`? I don't know what else to use, however I'm still ok with `class` personally. A Class doesn't have to be instantiated to exist as a Class, if it was instantiated I would call it `instance`. 

If that is still not good, then I don't know, maybe a word from this list: `accessor`, `builder`, `package`, `container`, `composer`, `template`, `member`, `metaclass`, `mixin`, `prototype`

**Comment by jmzagorski** [comment]
The only other one I could think of was `ctor` (for constructor). `class` is a reserved es6 keyword so I don't know if that is a good idea. `classReference` is good, but we might want to document that they dont have to use an actual es6 class and can use a plain old JavaScript function constructor like in the slickgrid editors file. `model` is very appealing too since it is a shorter name and it goes with your naming convention. By that i mean if a user didnt know what `editor.model` was they might figure it out since you have an editor interface file in the models folder? Lol idk. I think I actually like that one the best.

I won't be able to work on this or get on github until later today so let me know what you decide and I can start the changes later.

**Comment by ghiscoding** [comment]
Let's go with `model` then, I didn't want to use `type` because I want to make it obvious that both Filters and Editors are used differently compare to `1.x`. 

On my side, I think that I'm almost done with Grid State & Presets for all states (filter, sorter, pagination, column visibility, size & position). I have to test with the backend, which I'll do tomorrow. However, this branch is based on the new major 2 branch, so I need your code done before I can merge it anywhere.

Tomorrow is ok too, if you can finish it during the day. I'd like to make a Beta Monday or Tuesday night. That is if we are finished and are happy with the changes. Also the [Migration Guide](https://github.com/ghiscoding/aurelia-slickgrid/wiki/Migration-to-2.x) will have to be updated with your latest change.

**Comment by jmzagorski** [comment]
On the [compoundDateFilter](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/filters/compoundDateFilter.ts#L28) and [compoundInputFilter](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/filters/compoundInputFilter.ts#L27) you already have an operator property. Do you want me to use this property to set the default filter value instead of `defaultOperator`?

**Comment by ghiscoding** [comment]
Yeah it's normal for a Compound Filter to have an `operator` since it's dynamic and can change, so that one is the correct name. For the question, should we use `operator` everywhere? Hmm I don't have a preference, but I guess to make them all look the same, it's probably better to use `operator` as well indeed.  I'll leave it up to you.

**Comment by jmzagorski** [comment]
I was testing the filters with my changes and noticed a possible bug on exampe5. If I filter `a*=e` and then select either male or female, the grid returns no data and I do not see any exceptions. 

However on this branch i get 

`Unhandled rejection TypeError: Cannot read property 'toLowerCase' of undefined`

[here](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/examples/slickgrid/example5.ts#L206) because the `columnId` is "and gender", which does not exist in the column array (only `gender` does)

**Comment by ghiscoding** [comment]
That's ok, Example 5 is displaying data but I had to do a lot of work to fake an OData sample. It is not representing the reality of an OData Server. However, we can fix this easily by moving the `column[columnId]` right after the [for](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/examples/slickgrid/example5.ts#L200) loop and use a variable to keep it as a string `const col = column[columnId] + '';` and replace all the return with that variable. We now know it's a string, so we can call the `toLowerCase()` without exception. 

By the way, why would you filter `a*=e`? is that with a Compound Filter of `a*` and a search value of `=e`? That's like having 2 operators, which one will win?

**Comment by jmzagorski** [comment]
sorry, i meant `a*` and `e`. If it is just in the example, that is fine then.

**Comment by ghiscoding** [comment]
yeah it's just an example, it's kinda hard to repro a complete OData Server lol.
but that filter you've put, shouldn't throw an exception though. Unless you have `null` data in that column?

**Comment by jmzagorski** [comment]
if you put  `a*` filter with a value of `e` and then select a gender, then it throws

**Comment by ghiscoding** [comment]
Ok I made a quick patch, you can replace the `for` loop with this
```javascript
for (const columnId in columnFilters) {
  if (columnFilters.hasOwnProperty(columnId)) {
    filteredData = filteredData.filter(column => {
      const filterType = columnFilters[columnId].type;
      const searchTerm = columnFilters[columnId].term;
      let colId = columnId;
      if (columnId && columnId.indexOf(' ') !== -1) {
        const splitIds = columnId.split(' ');
        colId = splitIds[splitIds.length - 1];
      }
      const filterTerm = column[colId];
      if (filterTerm) {
        switch (filterType) {
          case 'equal': return filterTerm.toLowerCase() === searchTerm;
          case 'ends': return filterTerm.toLowerCase().endsWith(searchTerm);
          case 'starts': return filterTerm.toLowerCase().startsWith(searchTerm);
          case 'substring': return filterTerm.toLowerCase().includes(searchTerm);
        }
      }
    });
  }
}
```
This example never worked with multiple filters, now it does :)

**Comment by jmzagorski** [comment]
So I am done with my changes, but I am making sure I did not break anything because I had to make a lot of changes. However, I am noticing one difference between this branch and master. In example 4 when you clear all filters and then filter on the `complete` column with [for example] > 50 and then you remove the `>` and 50 the grid does not show all records again until you click "Clear all filter". Do you see this as well?

**Comment by ghiscoding** [comment]
It's because of the default operator that is set in the filter condition [here](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/filter-conditions/numberFilterCondition.ts#L8), if no operator is passed, it will use `==` which I thought would be the correct behavior if you start typing a number, you would expect it to be equal. 

However, this might be more confusing, so you can add this before the return
```javascript
if (!options.operator || options.operator === '') {
    return true;
}
```
but I'm not sure if that is the correct behavior either. The `testFilterCondition` doesn't really have a contain (it has `IN` but that is for a string contain). If you put 50 without an operator, will all the user expect to see nothing filtered? Probably not.

**Comment by jmzagorski** [comment]
At the end of the GIF when I clear 50 and the operator, shouldn't that clear the filters so I can see all the data? (I should have hit clear filters again in the GIF to show that all records were not shows when I cleared the operator and 50.)

**Comment by ghiscoding** [comment]
I just saw your GIF, it's a bit different than I understood. Yes it should put back all the data and it does in my current branch of the Grid State Columns, however there's a lot of fixes related to the filters, sorters in that branch. So I would say, push your code in and I will try to merge the 2 branches together (ideally make a PR to merge the new grid state branch into the major branch, then 1 final PR to include all of it)

**Comment by jmzagorski** [comment]
okay. I have a lot of changes in mine too, do you want me to create a branch so you can look it over? I did some things I was not sure about. Or do you want me to commit on this branch?

**Comment by ghiscoding** [comment]
Actually, it only works after I did the change. Can you update the `numberFilterCondition` with this code

```diff
export const numberFilterCondition: FilterCondition = (options: FilterConditionOption) => {
  const cellValue = parseFloat(options.cellValue);
  let searchTerm = (Array.isArray(options.searchTerms) && options.searchTerms[0]) || 0;
  if (typeof searchTerm === 'string') {
    searchTerm = parseFloat(searchTerm);
  }

+  if (!searchTerm && (!options.operator || options.operator === '')) {
+    return true;
+  }
  return testFilterCondition(options.operator || '==', cellValue, searchTerm);
};
```

Try it out and see if that is the behavior you want, you can tweak it if necessary

**Comment by ghiscoding** [comment]
@jmzagorski 
did you work on the same branch `new-major-2-release`? If so, I can take a look at the commits, hopefully less than 10 :P

Or if it's a different branch, you can create a PR to merge on the major branch

**Comment by jmzagorski** [comment]
I committed it all in one commit on this branch so we can revert it if you do not like it.

The one thing I was not sure about was my change in the `graphql.service` which removed the `mapOperatorByFilterType` [here](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/services/graphql.service.ts#L402). I tested on example 6 and everthing seemed to work because the `columnDef.filter.operator` always had the correct operator.

The one thing I thought about was forcing transient registration in the FilterFactory because if a user does not register their filter as Transient, it will default as singleton and probably not act right. However, I left this out in case there is an edge case where someone wants singleton.

**Side note**:  I am sure you are aware of this, but the date filters on example 12 do not filter based on a similar format filter value (see gif below), probably because of the underlying value being a date object vs string filter value. Just thought I would mention it while I was looking it over:

![2018-06-04_15-59-07](https://user-images.githubusercontent.com/8404957/40938574-3c9cd6f2-6810-11e8-8360-a0b1031d9542.gif)

**Comment by ghiscoding** [comment]
Example 12 should also use the Compound Date Filter, I can update that later. Technically what happens is that when you add the "-", it becomes an invalid date and that is why there's nothing returned

**Comment by ghiscoding** [comment]
@jmzagorski 
That was an amazingly long day, it took me couple of hours to release all of this and it seems that NPM has some issues showing my Beta version. I had similar issues on my work computer, so I guess I will wait few hours to see. NPM says that it got publish for version [2.0.0](https://www.npmjs.com/package/aurelia-slickgrid) with a Beta tag but I still don't see it

Anyhow, here's the last info on the project
1. merged both branches `grid-state-columns` and `major-release-2`
2. pushed to `master`
3. updated all `client-cli` and `github/doc`
   - `client-cli` throws an error with `flatpickr` when switching locale, e.g. on Example 12, click on "switch language" and it throws this error:
```bash
Error: Module name "flatpickr/dist/l10n/fr.js" has not been loaded yet for context: _
```
the error comes from this [line](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/filters/compoundDateFilter.ts#L222). Was the error there before? It's possible, but somehow this require doesn't work with loading the locale with `RequireJS` 
4. updated GitHub demo page and also added Example 15 with Grid State & Presets with Local Storage

There's only 1 major thing missing and it's to update the [Migration Guide](https://github.com/ghiscoding/aurelia-slickgrid/wiki/Migration-to-2.x) with the new changes you've put in for Editors/Filters. The rest should be covered.

**Comment by jmzagorski** [comment]
Great job! As for the flatpickr error I never ran into it, but I only use English (however my app will eventually have to be translated into many different languages.)

Do you want me to work on the migration guide for the editor/filter changes I committed?

**Comment by ghiscoding** [comment]
@jmzagorski 
Yes the migration is basically only missing the Editors/Filters changes you made yesterday. The rest should be covered and is well documented, I think

**Comment by jmzagorski** [comment]
I will work on that. My plan this week is to work on the migration documentation, fix all the unit tests in my app  to make sure everything works (my app was also in the process of going from V1 (Kendoui grid) to V2 (your grid)). Then I will close out our long documentation issue and integrate AureliaSlickgrid V2-Beta to see if anything works differently than expected.

**Comment by ghiscoding** [comment]
Sweet that is a nice plan :)
Hope our lib is better than KendoUI hehe

**Comment by ghiscoding** [comment]
@jmzagorski 
I found a new bug that annoys the hell out of me, I spent already half a day on it and I just can't figure out what causes it. The problem is that even if I clear (or change sort) and then I clear of change a filter, sometimes the original sorting comes back!?! I can always reproduce it if I go on Example 4, do a Clear Sorting and right after a Clear Filters, the sorting comes back, wth! From the investigation so far, I found that it comes from this [line](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/services/filter.service.ts#L130) `dataView.refresh()` inside the filter callback. Also it puts back the sort icon, only after reaching the callback of that column (basically from the sample below, after triggering a clear filter, it will start showing the sort icon only after reaching the "Duration" column). I have a feeling that it's because of a reference pointer and possibly since our refactoring to support multiple grid. 

I don't have an older version anymore, so I can't really test this. Or perhaps, I could use the Github demo and use previous Aurelia-Slickgrid version but I would have to remodify the example 4 again. 


![2018-06-05_16-25-48](https://user-images.githubusercontent.com/643976/41000884-6b17ec24-68dd-11e8-87ef-39ca308d8da3.gif)

**Comment by jmzagorski** [comment]
Good thing I am behind on my pulls. My master is 9 commits behind, the last commit I have is `refactor(demo): update to latest version 1.13.0` and this problem does occur so it is not in our new stuff

**Comment by jmzagorski** [comment]
Found the issue: the clear filters calls `dataView.onRowCountChanged` which the `sort.service` subscribes to [here](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/services/sort.service.ts#L128). The `args.current` from that event call is 478, which is greater than 0 so the local presets are loaded on the grid.

**Comment by ghiscoding** [comment]
oh ok good thing I asked then, did you just find the issue or do you know how to fix it?

**Comment by ghiscoding** [comment]
@jmzagorski 
I'm refactoring the Editors to support validators. I didn't know until today that we can watch `grid.onValidationError` to deal with the result of the `validate()` function in the Editors and display that validation to the end user. 

My question to you is the following, can we pass a custom `validator` to the `editor` property in the Grid Definitions? I doesn't work in my Angular repo because I didn't refactor the code that you've done recently with `model` in `editor` and `filter`, but would that work in Aurelia-Slickgrid?

Basically a custom validator would look like this
```javascript
const myCustomTitleValidator = (value) => {
  if (value == null || value === undefined || !value.length) {
    return { valid: false, msg: 'This is a required field' };
  } else {
    return { valid: true, msg: null };
  }
};
```
and pass it to the definitions
```javascript
this.columnDefinitions = [
  { id: 'title', field: 'title', editor: { model: Editors.longText }, validator: myCustomTitleValidator }
];
```

I didn't try it since I'm not in Aurelia, but I wish it works. Can you confirm?

**Comment by ghiscoding** [comment]
@jmzagorski 
not sure if it's really a good topic or not but I was looking at the implementation of `Editors` and saw the `factory.of` on this [line](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/aurelia-slickgrid.ts#L157). This makes me thing, we could possibly do that also with the `formatters` to have DI working and get rid of the `i18n` in the grid options. I think it's only `Formatters` that aren't DI yet. For example the `translateFormatter` needs the `i18n` from grid options, as shown on this [line](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/formatters/translateFormatter.ts#L6) but if we had DI that would fix that formatter.

**Comment by jmzagorski** [comment]
Right, formatters are more challenging. Two methods I can think of off the top of my head right now:

1. You would have to define an interface for the user to use so your library knew what function to call on the class after it was instantiated
2. Or the user would have to create a class with their dependencies as the first parameters to their class constructor and then `factory.of` feeds the rest of the arguments to the constructor, similar to the editor. Then when slickgrid runs it will be calling this Aurelia's [invoke](https://github.com/aurelia/dependency-injection/blob/master/src/container.js#L449) every time (not sure what type of overhead that brings)
For example:
```javascript
@inject(i18n)
export class MyFormatter {
  constructor(i8n i8n, row, cell, value, columnDef, dataContext) {
  }
}
```

I am sure there are better ways and I can think of some as I am going through my code.

**Comment by jmzagorski** [comment]
what if you move the creation of the filter [here](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/services/filter.service.ts#L371) up higher in the code to [here](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/services/filter.service.ts#L351) so the operator logic can be

 `operator = columnDef.filter.operator || filter.operator;`

Then you can either pass it to `updateColumnFilters` or add it to the `columnDef.filter.operator`.

**Comment by ghiscoding** [comment]
@jmzagorski 
Yeah that is a lot better and it simplifies the code too, thanks.

**Comment by ghiscoding** [comment]
@jmzagorski 
oops talked too fast, Example 4 was ok but Example 12 broke since `columnDef.filter` is undefined. I changed your line to this
```javascript
operator = filter.operator || (columnDef && columnDef.filter && columnDef.filter.operator);
```
is that ok?

**Comment by ghiscoding** [comment]
Yeah I see that in the in `ColumnFilter`, the `operator` can be used and maybe defined in the column definition, so I will flip it. Also a `filter` can be `undefined`, so I have to take that in consideration too. 

So the final result is this

```diff
addFilterTemplateToHeaderRow(args: { column: Column; grid: any; node: any }) {
    const columnDef = args.column;
    const columnId = columnDef.id || '';

    if (columnDef && columnId !== 'selector' && columnDef.filterable) {
      let searchTerms: SearchTerm[] | undefined;
      let operator: OperatorString | OperatorType;
+      const filter: Filter | undefined = this.filterFactory.createFilter(args.column.filter);
+      operator = (columnDef && columnDef.filter && columnDef.filter.operator) || (filter && filter.operator) || undefined;

      if (this._columnFilters[columnDef.id]) {
        searchTerms = this._columnFilters[columnDef.id].searchTerms || undefined;
        operator = this._columnFilters[columnDef.id].operator || undefined;
      } else if (columnDef.filter) {
        // when hiding/showing (with Column Picker or Grid Menu), it will try to re-create yet again the filters (since SlickGrid does a re-render)
        // because of that we need to first get searchTerm(s) from the columnFilters (that is what the user last entered)
        searchTerms = columnDef.filter.searchTerms || undefined;
-        this.updateColumnFilters(searchTerms, columnDef);
-        operator = columnDef.filter.operator || undefined;
+        this.updateColumnFilters(searchTerms, columnDef, operator);
      }
```

I finally took the time to port it over to my Angular repo, which is how I find this issue.

**Comment by jmzagorski** [comment]
do you need the last `|| undefined` because of typescript?

**Comment by ghiscoding** [comment]
well I just saw that the `filter` is `const filter: Filter | undefined` so if that is the case and it can't find the `operator` in `columnDef`, neither in `filter`, then it will be that 3rd value returned. TS didn't give me any error in my Angular repo but the TSlint is not as strict as it is in my Aurelia repo. I'm quite sure I'll have to do that in Aurelia because of the `filter` that can be `undefined` (which will actually never happen but it is declared that way so)

**Comment by jmzagorski** [comment]
if your first or second expression is false then it will be undefined without the 3rd expression because there will be no 3rd expression to evaluate.

For example if `columnDef = undefined` and `filter = undefined` than the result will be `undefined` with this expression too `(columnDef && columnDef.filter && columnDef.filter.operator) || (filter && filter.operator)`. You can leave it as you have it since the result will be the same, but just thought I would mention that.

**Comment by ghiscoding** [comment]
oh right I've been working on this for 2 late nights recently and don't see these obvious things anymore lol

**Comment by jmzagorski** [comment]
lol I have been there too so I know what you are talking about.

**Comment by ghiscoding** [comment]
@jmzagorski 
I have updated the Migration Guide, since I had to do the Angular one too.

**Comment by ghiscoding** [comment]
For reference, the BUG mentioned [here](https://github.com/ghiscoding/aurelia-slickgrid/issues/19#issuecomment-395523294) is fixed with this [commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/eacd6337ee27fe836644fd530577b3e3596822b3)

The Aurelia-Slickgrid `2.x` final will be released early next week. I plan to add Slider Editor this weekend and couple of other changes before being ready for final version.

**Comment by jmzagorski** [comment]
thanks for updating the migration guide. It looks like the changes I did are in there as well. Sorry I have not been able to do anything. We are missing a few developers so I have to do some extra work over the past week and a half.
